tcti/tcti_socket.c

```diff
@@ -528,3 +534,65 @@ TSS2_RC InitSocketTcti (
 
     return rval;
 }
+
+TSS2_RC Tss2_Tcti_Socket_Init (
+    TSS2_TCTI_CONTEXT *tctiContext,
+    size_t *size,
+    const char *conf
+    )
+{
+    TCTI_SOCKET_CONF sock_conf = { 0 };
+    TSS2_RC rc;
+    UriParserStateA state;
+    UriUriA uri;
+    const char *uri_str = conf != NULL ? conf : TCTI_SOCKET_DEFAULT;
+    size_t range;
+    /* maximum 5 digits in uint16_t + 1 for \0 */
+    char port[6] = { 0 };
+    char hostname[HOST_NAME_MAX + 1] = { 0 };
+
+    state.uri = &uri;
+    if (uriParseUriA (&state, uri_str) != URI_SUCCESS) {
+        rc = TSS2_TCTI_RC_BAD_VALUE;
+        goto out;
+    }
+
+    /* extract host & domain name / fqdn */
+    range = uri.hostText.afterLast - uri.hostText.first;
+    if (range > HOST_NAME_MAX) {
+        rc = TSS2_TCTI_RC_BAD_VALUE;
+        goto out;
+    }
+    strncpy (hostname, uri.hostText.first, range);
+    sock_conf.hostname = hostname;
+
+    /* extract port number */
+    range = uri.portText.afterLast - uri.portText.first;
+    if (range < 5 && range > 0) {
+        strncpy (port, uri.portText.first, range);
+        sscanf (port, "%" SCNu16, &sock_conf.port);
+    } else if (range == 0) {
+        sock_conf.port = DEFAULT_SIMULATOR_TPM_PORT;
+    } else { /* range > 5 */
+        rc = TSS2_TCTI_RC_BAD_VALUE;
+        goto out;
+    }
+    rc = InitSocketTcti (tctiContext, size, &sock_conf, 0);
+out:
+    uriFreeUriMembersA (&uri);
+    return rc;
+}
+
+/* public info structure */
+static TSS2_TCTI_INFO tss2_tcti_info __attribute__ ((unused)) = {
+    .name = "tcti-socket",
+    .description = "TCTI module for communication with the Microsoft TPM2 Simulator.",
```

Comment:
Is that only MS TPM2 Simulator? What about a TPM on a remote platform (which is not yet supported)

Reply:
The socket TCTI implements a protocol on top of the normal TPM2 command / response messages. This protocol was defined by Microsoft and published as part of their TPM2 reference code / implementation that's in the TCG specs. You can use this TCTi to communicate with a local simulator or one on a remote system currently (not sure if we consider this "supported").

You could use this same protocol to communicate with any TPM so long as some "middle-ware" implements the protocol on top of the underlying TPM (this is what we did in the old `resourcemgr`) but I wouldn't recommend it. I'm thinking of trying out something with protocol buffers in the tabrmd but haven't started: https://github.com/intel/tpm2-abrmd/issues/138